specs/lazy-adr/adr-018-forced-inclusion-mechanism.md

```diff
@@ -0,0 +1,462 @@
+    DAOriginHash      []byte         // Hash of the DA layer block this rollup block derives from
+    DAOriginTimestamp uint64         // Timestamp of the DA layer block
+    DirectTxsIncluded []common.Hash  // Hashes of direct transactions included from the DA layer
+    // Other existing extraData fields, such as sequencer information
+}
+```
+
+This approach maintains compatibility with the minimal header structure while adding the necessary information for forced inclusion.
+
+#### Direct Transaction Tracking
+
+```go
+type DirectTransaction struct {
+    TxHash       common.Hash  // Hash of the transaction
+    DAHeight     uint64       // Height at which it was included in the DA layer
+    DAHash       []byte       // Hash of the DA block
+    Included     bool         // Whether it has been included in a rollup block
+    IncludedAt   uint64       // Height at which it was included in the rollup
+}
+
+type DirectTxTracker struct {
+    txs         map[common.Hash]DirectTransaction  // Map of direct transactions
+    mu          sync.RWMutex                       // Mutex for thread-safe access
+    latestDAHeight uint64                          // Latest DA block height scanned
+}
+```
+
+#### Sequencer Status Tracking
+
+```go
+type SequencerStatus struct {
+    IsActive          bool      // Whether the sequencer is considered active
+    LastActiveHeight  uint64    // Last DA height where sequencer posted a batch
+    DownSince         time.Time // Time since sequencer was last seen active
+}
+```
+
+### APIs and Interfaces
+
+#### Enhanced DA Client Interface
+
+```go
+type DAClient interface {
+    // Existing methods
+    // ...
+    
+    // New method for forced inclusion
+    GetDirectTransactions(ctx context.Context, fromHeight, toHeight uint64) ([][]byte, error)
+    // Note: SubmitDirectTransaction is removed as it's not a responsibility of the rollup node
+}
+```
+
+#### Sequencer Interface Extensions
+
+```go
+// New methods added to the Sequencer interface
+func (s *Sequencer) ScanDALayerForDirectTxs(ctx context.Context) error
+func (s *Sequencer) IncludeDirectTransactions(ctx context.Context, batch *Batch) error
+func (s *Sequencer) ValidateDAReference(ctx context.Context, header *types.Header) error
+```
+
+#### Full Node Interface Extensions
+
+```go
+// New methods added to the Node interface
+func (n *Node) CheckSequencerStatus(ctx context.Context) (bool, error)
+func (n *Node) ProcessDirectTransactions(ctx context.Context) error
+func (n *Node) ValidateBlockWithDAReference(ctx context.Context, block *types.Block) error
+```
+
+### Implementation Changes
+
+#### Single Sequencer Node Changes
+
+1. **DA Layer Scanner**:
+   - Implement a periodic scanner that queries the DA layer for direct transactions
+   - Track all direct transactions in the DirectTxTracker data structure
+   - Update the latest scanned DA height after each scan
```

Comment:
this seems like an expensive query. will we need to scan the entire namespace and compare to what we have locally?